Java: Views and Projections (#1916)
cap-js/docs: java/working-with-cql/query-execution.md

long deleteCount = service.run(delete).rowCount();
```

## Views and Projections { #views }

The CDS compiler automatically generates [DDL](../../guides/databases?impl-variant=java#generating-sql-ddl) files from your CDS model, including database views for all CDS [views and projections](../../cds/cdl#views-projections).

These views are used to read data from the database, while write operations are resolved to the underlying entities, if possible.

> ::: warning

Author: ```suggestion
::: warning Avoid to-many associations
```

Avoid using to-many associations on the select list of CDS views, as this prevents write operations and can lead to performance issues due to duplicating records on the DB.
:::

### Runtime Views { #runtimeviews }

To add or update CDS views without redeploying the database schema, annotate your views with [@cds.persistence.skip](../../guides/databases#cds-persistence-skip). This tells the CDS compiler to skip generating database views for these entities. Instead, the CAP Java runtime dynamically resolves these views at runtime.

::: warning Limitations
Runtime views do not support complex views that use aggregations, unions, joins, or subqueries in the `FROM` clause.

> To read [draft-enabled](../fiori-drafts#reading-drafts) entities, set `cds.drafts.persistence` to `split`.

Reviewer: Would be nice to explain that drafts are not runtime views and still remain the plain tables. This confused some stakeholders: e.g. that value from path expression is writable through draft and not writable through view.

```

::: warning Limitations of `resolve` mode
Expands to other runtime views and complex draft queries are not supported.
:::

### Write through Views { #updatable-views}

To run [Insert](./query-api#insert), [Upsert](./query-api#upsert), [Update](./query-api#update), and [Delete](./query-api#delete) statements on CDS views, the CAP Java runtime resolves them to the underlying entity definitions, if possible - similar to the runtime view *resolve* mode for queries.

When delegating queries between Application Services and Remote Services, statements are also resolved to the entity definitions of the targeted service.

Views and projections can be resolved for write operations if the following conditions are met:

- The view definition does not use any other clause than `columns` and `excluding` - `join`, `union` and `where` are not supported.
- The projection includes all key elements; with the exception of insert operations with generated UUID keys.
- The projection includes all elements with a `not null` constraint, unless they have a default value.
- The projection must not include [path expressions](../../cds/cql#path-expressions) using to-many associations.

> - Projections targeting remote OData services must not include calculated elements.

Reviewer: Can view include mixin associations or compositions? What will be writable then?

For [Insert](./query-api#insert) or [Update](./query-api#update) operations, if the projection contains functions or expressions, these values are ignored. 

Path expressions navigating *to-one* compositions, can be used in projections as shown by `headerStatus` in the following example, which includes the element `status` of the `OrderHeader` entity.

```cds
entity Order as projection on bookshop.Order;
entity Order as projection on bookshop.Order { ID, status as state };

entity Order as projection on bookshop.Order {
  ID,
  header.status        as headerStatus,
  header.customer.name as customerName @readonly
} excluding { status };
```

> ::: warning Path Expressions using Associations are readonly

Author: ```suggestion
::: warning Path Expressions using associations are readonly
```

Path expressions navigating associations are by default [not writable](#cascading-over-associations) and need to be annotated with [@readonly](../../guides/providing-services#readonly).
:::

> Write operations on views that cannot be resolved by the CAP Java runtime remain unmodified

Reviewer: Maybe also cross-reference that drafts must be handled through an application service. Just to make people aware of this one more time.

long deleteCount = service.run(delete).rowCount();
```

## Views and Projections { #views }

The CDS compiler generates [DDL](../../guides/databases?impl-variant=java#generating-sql-ddl) files from your CDS model, including database views for all CDS [views and projections](../../cds/cdl#views-projections).

Entity tables and views are deployed to the database. Read statements on CDS views target the database view, while write operations are resolved to the underlying entity table by the CAP Java runtime, when possible.

::: warning Avoid to-many associations
Do not use to-many associations in the select clause of CDS views. This blocks write operations and can cause performance issues due to record duplication.
:::

### Runtime Views { #runtimeviews }

To add or update CDS views without redeploying the database schema, annotate your views with [@cds.persistence.skip](../../guides/databases#cds-persistence-skip). This tells the CDS compiler to skip generating database views for these entities, and the CAP Java runtime resolves them dynamically at runtime.

::: warning Limitations

> Runtime views do not support aggregations, unions, joins, or subqueries in the FROM clause.

Reviewer: Not sure if we need that many warning everywhere.

Author: I'll change this to an info box

  key ID     : UUID;
      title  : String;
      stock  : Integer;
      author : Association to one Authors;
}
@cds.persistence.skip
entity BooksWithLowStock as projection on Books {
    ID, title, author.name as author
} where stock < 10; // makes the view read only
```
```sql
SELECT from BooksWithLowStock where author = 'Kafka'
```

CAP Java supports two modes for resolving runtime views:

**`cte` mode**: The runtime translates the view definition into a _Common Table Expression_ (CTE) and sends it with the query to the database.

```sql
WITH BOOKSWITHLOWSTOCK_CTE AS (
    SELECT B.ID,
           B.TITLE,
           A.NAME AS "AUTHOR"
      FROM BOOKS B
      LEFT OUTER JOIN AUTHOR A ON B.AUTHOR_ID = A.ID
     WHERE B.STOCK < 10
)
SELECT ID, TITLE, AUTHOR AS "author"
  FROM BOOKSWITHLOWSTOCK_CTE
 WHERE A.NAME = ?
```

::: tip
In CAP Java 3.10, enable `cte` mode with **cds.sql.runtimeView.mode: cte**
:::

::: warning Draft-enabling runtime views requires schema update
When [draft-enabling](../fiori-drafts#reading-drafts) a runtime view, a corresponding draft persistence table is created for the CDS view and a database schema update is required when changing the runtime view. 

> [Draft activate](../fiori-drafts#editing-drafts) triggers a [write through view](#updatable-views) and the restrictions apply.

Reviewer: Restrictions of write-through? 

I would rather gather draft-related part in the separate subsection so that later it can be referenced.

:::

**`resolve` mode**: The runtime _resolves_ the view definition to the underlying persistence entities and executes the query directly against the corresponding tables.

```sql
SELECT B.ID, B.TITLE, A.NAME AS "author"
  FROM BOOKS AS B
  LEFT OUTER JOIN AUTHORS AS A ON B.AUTHOR_ID = A.ID
 WHERE B.STOCK < 10 AND A.NAME = ?
```

::: warning Limitations of `resolve` mode
Using associations introduced by the view (mixins), as well as complex draft queries, and [draft-enabling](../fiori-drafts#reading-drafts) runtime views are not supported in *resolve* mode.
:::

> 

Reviewer: Maybe also add that runtime views are not lockable. See https://cap.cloud.sap/docs/java/working-with-cql/query-api#write-lock

Author: You're right, HANA seems to not support locks via CTEs, so only the resolve mode works. I'll add a note.

Reviewer: Maybe we should fix this? E.g. use different mode for queries with locks? They are usually very simple and should resolve to the baseline entity 1:1.

When delegating queries between Application Services and Remote Services, statements are also resolved to the targeted service's entity definitions.

Write operations on CDS views are supported if:

- The view definition uses only *columns* and *excluding* (no *join*, *union*, or *where*).
- The projection includes all *not null* elements (incl. keys), unless they have a default or generated value.
- The projection does not include [path expressions](../../cds/cql#path-expressions) using *to-many* associations.
- Projections targeting *remote OData* services must not include calculated elements.

You can use path expressions navigating *to-one* compositions, as shown by `headerStatus`:

```cds
// CDS views supporting write operations
entity Order as projection on bookshop.Order excluding { status };

entity Order as projection on bookshop.Order {
  key ID,
      header.status        as headerStatus,
      header.customer.name as customerName @readonly,
      items                as lineItems // aliased composition
};
```

> [Deep write](./query-execution#deep-insert-upsert) via (aliased) compositions is supported if there are corresponding compositions in the underlying entity definition. Deep write via compositions that are only defined in the view (mixins) is not supported and the data for such mixin compositions is ignored. Data for elements corresponding to *expressions* and *functions* is ignored as well.

Reviewer: >Data for elements corresponding to *expressions* and *functions* is ignored as well.

Sample? 

Also, why not merge this in the list above? It can be nicer to have them listed as a reference.

long deleteCount = service.run(delete).rowCount();
```

## Views and Projections { #views }

The CDS compiler generates [DDL](../../guides/databases?impl-variant=java#generating-sql-ddl) files from your CDS model, including database views for all CDS [views and projections](../../cds/cdl#views-projections).

Entity tables and views are deployed to the database. Read statements on CDS views target the database view, while write operations are resolved to the underlying entity table by the CAP Java runtime, when possible.

> ::: tip Avoid mixin compositions

Reviewer: This seems to be sufficiently explain later. Why mention it here?

> ::: tip Avoid mixin compositions

Author: These are independent of the type of view and should go into a general section

```cds
entity Books {
  key ID     : UUID;
      title  : String;
      stock  : Integer;
      author : Association to one Authors;
}
@cds.persistence.skip
entity BooksWithLowStock as projection on Books {
    ID, title, author.name as author
} where stock < 10; // makes the view read only
```
```sql
SELECT from BooksWithLowStock where author = 'Kafka'
```

CAP Java supports two modes for resolving runtime views:

#### CTE mode

This is the default mode in CAP Java `4.x`. The runtime translates the view definition into a _Common Table Expression_ (CTE) and sends it with the query to the database.

```sql
WITH BOOKSWITHLOWSTOCK_CTE AS (
    SELECT B.ID,
           B.TITLE,
           A.NAME AS "AUTHOR"
      FROM BOOKS B
      LEFT OUTER JOIN AUTHOR A ON B.AUTHOR_ID = A.ID
     WHERE B.STOCK < 10
)
SELECT ID, TITLE, AUTHOR AS "author"
  FROM BOOKSWITHLOWSTOCK_CTE
 WHERE A.NAME = ?
```

::: tip CAP Java 3.10
Enable *cte* mode with *cds.sql.runtimeView.mode: cte*
:::

#### Resolve mode

The runtime _resolves_ the view definition to the underlying persistence entities and executes the query directly against the corresponding tables.

> ```sql

Reviewer: The more I read this (especially these references later in chapters on modifications and deletions), the more I feel that we are to vague about what this `resolve` mode actually does.

Reviewer: Maybe we should not reference this mode later for modifications. This is actually very hard to match the statements: 
- this mode does not support draft-enabled entities (for read, but this never clearly stated)
- the updates write through view using similar technique. 
- but drafts somehow function? What exactly will work?

Author: Restructured the section and removed the refs

SELECT B.ID, B.TITLE, A.NAME AS "author"
  FROM BOOKS AS B
  LEFT OUTER JOIN AUTHORS AS A ON B.AUTHOR_ID = A.ID
 WHERE B.STOCK < 10 AND A.NAME = ?
```

::: info Limitations of `resolve` mode
Using associations introduced by the view (mixins), as well as complex draft queries, and [draft-enabling](../fiori-drafts#reading-drafts) runtime views are not supported in *resolve* mode.
:::

### Draft with Runtime Views { #draft-views }

If you define runtime views on [draft-enabled](../fiori-drafts#reading-drafts) entities and want to run draft specific queries on these views, set the *cds.drafts.persistence* configuration to `split` and run the queries through the [Draft Service](../fiori-drafts#draft-service) or [Application Service](../cqn-services/application-services#application-services). The [Persistence Service](../cqn-services/persistence-services) only works for non-draft specific queries.

::: warning Avoid draft-enabling runtime views

> [Draft-enabling](../fiori-drafts#reading-drafts) runtime views is only supported in *cte* mode. The cds-compiler creates a corresponding draft persistence table for the view and a database schema update is required each time the runtime view is changed.

Reviewer: Maybe this whole chapter should sound like: if you draft-enable runtime views make sure that `CTE` mode is used and draft persistence is in `split` mode, you should use draft service for queries to them (which is general requirement) and your view is updatable. Like this whole chapter is a warning without the yellow box.

>cds-compiler

CDS Compiler?

Write operations on CDS views are supported if:

For [Insert](./query-api#insert) or [Update](./query-api#update) operations, if the projection contains functions or expressions, these values are ignored. Path expressions navigating *to-one* associations, can be used in projections as shown by the `Header` view in the following example. The `Header` view includes the element `country` from the associated entity `Address`.
- The view definition uses only *columns* and *excluding* (no *join*, *union*, or *where*).
- The projection includes all *not null* elements (incl. keys), unless they have a default or generated value.
- The projection does not include [path expressions](../../cds/cql#path-expressions) using *to-many* associations.
- Projections targeting *remote OData* services must not include calculated elements.

You can use path expressions navigating *to-one* compositions, as shown by `headerStatus`:

```cds
// Supported
entity Order as projection on bookshop.Order;
entity Order as projection on bookshop.Order { ID, status as state };
// CDS views supporting write operations
entity Order as projection on bookshop.Order excluding { status };

> entity Header as projection on bookshop.OrderHeader { key ID, address.country as country };

Reviewer: This model should have reference to its base model. It is not clear that `header` is a composition.

::: warning Path Expressions in Views

> Path expressions navigating *associations* (e.g. *header.customer.name*) are [not writable](#cascading-over-associations) by default. To avoid issues on write, annotate them with [@readonly](../../guides/providing-services#readonly).

Reviewer: I think that warning is not necessary here. We just list instructions: the one about mixins (maybe also add a reference to some definition of this concept) is as important as the paths inside associations not being writable. 

Except the keys, right?

}
```
```sql
DELETE from OrderView where ID = 42
```
The delete operation is resolved to the underlying `Order` entity with ID *42* and cascades over the `header` and `items` compositions. The `delivery` composition, which is only defined in the view, is ignored and does not cascade the delete operation to `Delivery`.

### Runtime Views { #runtimeviews }

To add or update CDS views without redeploying the database schema, annotate them with [@cds.persistence.skip](../../guides/databases#cds-persistence-skip). This tells the CDS compiler to skip generating database views for these CDS views, and the CAP Java runtime resolves them dynamically on each request.

Runtime views must be simple [projections](../../cds/cdl#as-projection-on), not using *aggregations*, *join*, *union* or *subqueries* in the *from* clause, but may have a *where* condition if they are only used to read. On write, the restrictions for [write through views](#updatable-views) apply in the same way as for standard CDS views. However, if a runtime view cannot be resolved, a fallback to database views is not possible, and the statement fails with an error.

> **Read via Runtime Views** - consider the following CDS model and query:

Reviewer: Typo? Looks weird in preview.

Author: The formatting?

Reviewer 2: done

Author: Thanks :-)

```cds
entity Books {
  key ID     : UUID;
      title  : String;
      stock  : Integer;
      author : Association to one Authors;
}
@cds.persistence.skip
entity BooksWithLowStock as projection on Books {
    ID, title, author.name as author
} where stock < 10; // makes the view read only
```
```sql
SELECT from BooksWithLowStock where author = 'Kafka'
```

CAP Java supports two modes for resolving runtime views on read:

> #### Read in `CTE` mode { #rtview-cte }

Author: I'll change this to lower case

> #### Read in `CTE` mode { #rtview-cte }

Author: ```suggestion
#### Read in `cte` mode { #rtview-cte }
```

This is the default mode in CAP Java `4.x`. The runtime translates the view definition into a _Common Table Expression_ (CTE) and sends it with the query to the database.

```sql
WITH BOOKSWITHLOWSTOCK_CTE AS (
    SELECT B.ID,
           B.TITLE,
           A.NAME AS "AUTHOR"
      FROM BOOKS B
      LEFT OUTER JOIN AUTHOR A ON B.AUTHOR_ID = A.ID
     WHERE B.STOCK < 10
)
SELECT ID, TITLE, AUTHOR AS "author"
  FROM BOOKSWITHLOWSTOCK_CTE
 WHERE A.NAME = ?
```

::: info Limitations of `cte` mode
[Pessimistic locking](#pessimistic-locking) queries on runtime views in *cte* mode are not supported on SAP HANA.
:::

::: tip CAP Java 3.10
Enable *cte* mode with *cds.sql.runtimeView.mode: cte*
:::

> #### Read in `Resolve` mode { #rtview-resolve }

Reviewer 2: Is lowercase fine?
```suggestion
#### Read in `resolve` mode { #rtview-resolve }
```

```

## Views and Projections { #views }

The CDS compiler generates [DDL](../../guides/databases?impl-variant=java#generating-sql-ddl) files from your CDS model, by default including database views for all CDS [views and projections](../../cds/cdl#views-projections). These tables and views are deployed to the [database](../cqn-services/persistence-services#database-support). 

CDS views can be defined by simple [projections](../../cds/cdl#as-projection-on) or complex [selects](../../cds/cdl#as-select-from) including *joins*, *unions*, and *aggregations*. Read statements on CDS views target the generated database views.

For [write](#updatable-views) and [delete](#delete-via-view) operations, the CAP Java runtime attempts to resolve the CDS views to their underlying persistence entities, which is only supported for simple projections.

::: warning Avoid selecting to-many Associations
Do not use [*to-many associations*](../../cds/cdl#to-many-associations) in the select clause of CDS views. This blocks write operations and may cause performance issues due to record duplication on read.
:::
::: warning Avoid Composition Definitions in Views

> Avoid [defining](../../cds/cql#association-definitions) new *compositions* in CDS views and prefer *associations* instead, as [deep write](#updatable-views) and [cascading delete](#delete-via-view) are only supported for compositions in persistence entities.

Reviewer: Isn't this true only for runtime views? I seen stakeholders who used such models and that surprisingly worked.

      items                  as lineItems,        // writable composition
      toUpper(shipToCountry) as country : String  // ignored on write
};
```
```sql
UPDATE entity OrderView2
  { ID: 42, headerStatus: 'Processing', lineItems: [{ID: 1, book:251}] }
```
- Data for elements corresponding to *expressions* and *functions* (*country*) is ignored.
- [Deep write](./query-execution#deep-insert-upsert) via (aliased) compositions (*lineItems*) is supported if there are corresponding compositions (*items*) in the underlying entity definition. Deep write via compositions that are only defined in the view (for example via [mixins](../../cds/cql#association-definitions)) is not supported and the data is ignored.
- [Path expressions](../../cds/cql#path-expressions) over compositions (*header.status*) are writable. For [Inserts](./query-api#insert), the view must expose all *not null* elements of the target entity and the data must include values for all of them. In the example above, the order header must have a generated key to support inserting new orders with a value for *headerStatus*.

    ::: warning Path Expressions over Associations

>     Path expressions navigating *associations* (*header.customer.name*) are [not writable](#cascading-over-associations) by default. To avoid issues on write, annotate them with [@readonly](../../guides/providing-services#readonly).

Author: Above it's only a side note and since `@readonly` should be added, a box makes it clearer

:::

For example, the following CDS model defines `Order` with header and items, and `OrderView` which excludes header and exposes items as `lineItems`:

```cds
entity Order : cuid, managed {
      header : Composition of one OrderHeader;
      items  : Composition of many OrderItem on items.parent = $self;
}
entity OrderView as projection on db.Order {
  ID,
  items as lineItems,
  delivery : Composition of one Delivery on delivery.order = $self
}
```
```sql
DELETE from OrderView where ID = 42
```
The delete operation is resolved to the underlying `Order` entity with ID *42* and cascades over the `header` and `items` compositions. The `delivery` composition, which is only defined in the view, is ignored and does not cascade the delete operation to `Delivery`.

### Runtime Views { #runtimeviews }

To add or update CDS views without redeploying the database schema, annotate them with [@cds.persistence.skip](../../guides/databases#cds-persistence-skip). This tells the CDS compiler to skip generating database views for these CDS views, and the CAP Java runtime resolves them dynamically on each request.

> Runtime views must be simple [projections](../../cds/cdl#as-projection-on), not using *aggregations*, *join*, *union* or *subqueries* in the *from* clause, but may have a *where* condition if they are only used to read. On write, the restrictions for [write through views](#updatable-views) apply in the same way as for standard CDS views.

Reviewer 3: It might be a bit obvious but you could mention that the fallback on the database view does not work as there is none existing. Also, what would happen in case one runs into those restrictions? Does the request fail?

**Example** - consider the following CDS model and query:

```cds
entity Books {
  key ID     : UUID;
      title  : String;
      stock  : Integer;
      author : Association to one Authors;
}
@cds.persistence.skip
entity BooksWithLowStock as projection on Books {
    ID, title, author.name as author
} where stock < 10; // makes the view read only
```
```sql
SELECT from BooksWithLowStock where author = 'Kafka'
```

CAP Java supports two modes for resolving runtime views on read:

#### Read in `CTE` mode { #rtview-cte }

> This is the default mode in CAP Java `4.x`. The runtime translates the view definition into a _Common Table Expression_ (CTE) and sends it with the query to the database.

Reviewer 3: Love the new default! 😍

     WHERE B.STOCK < 10
)
SELECT ID, TITLE, AUTHOR AS "author"
  FROM BOOKSWITHLOWSTOCK_CTE
 WHERE A.NAME = ?
```

::: info Limitations of `cte` mode
[Pessimistic locking](#pessimistic-locking) queries on runtime views in *cte* mode are not supported on SAP HANA.
:::

::: tip CAP Java 3.10
Enable *cte* mode with *cds.sql.runtimeView.mode: cte*
:::

#### Read in `Resolve` mode { #rtview-resolve }

The runtime _resolves_ the view definition to the underlying persistence entities and executes the query directly against the corresponding tables.

```sql
SELECT B.ID, B.TITLE, A.NAME AS "author"
  FROM BOOKS AS B
  LEFT OUTER JOIN AUTHORS AS A ON B.AUTHOR_ID = A.ID
 WHERE B.STOCK < 10 AND A.NAME = ?
```

::: info Limitations of `resolve` mode
Using associations that are only [defined](../../cds/cql#association-definitions) in the view, as well as complex draft queries are not supported in *resolve* mode.
:::

### Draft Queries on Views { #draft-views }

When draft-enabling a CDS view, the CDS Compiler creates a corresponding draft persistence table for this view. [Draft activate](../fiori-drafts#editing-drafts) updates the active entity via the view, therefore

> - [Draft-enabled](../fiori-drafts#reading-drafts) CDS views must fulfill all requirements of [updatable views](#updatable-views).

Reviewer 2: Here's a suggestion:
![image](https://github.com/user-attachments/assets/8f566563-54ad-42d8-82e3-98ca84cde6ad)

long deleteCount = service.run(delete).rowCount();
```

## Views and Projections { #views }

With CDS [views](../../cds/cdl#views-projections) you can derive new entities from existing ones, for example to rename or exclude certain elements, or to add [virtual elements](../../cds/cdl#virtual-elements-in-views) for specific use cases.

> ::: tip Prefer simple views

Reviewer 4: s.u.

Prefer creating multiple simple views, each tailored to a specific use case, rather than a single complex view that tries to address multiple use cases simultaneously.
:::

The CDS compiler generates [DDL](../../guides/databases?impl-variant=java#generating-sql-ddl) files from your CDS model, including SQL views for the CDS views. These views are deployed to the [database](../cqn-services/persistence-services#database-support) and used by the CAP runtime to read data.

> For *read-only* views, you can use the full feature set of [selects](../../cds/cdl#as-select-from), including *joins*, *unions*, and *aggregations*. However, such complex views are not writable and require a schema redeployment if the view definition is changed.

Reviewer 4: ```suggestion
In addition, views that are used only for read access offer complex features to recombine data including *joins*, *unions*, and *aggregations*. However, be aware that such views enhanced with complex features **do not allow write  queries and hence are restricted to read-only access**. Also such views always require schema redeployment if the view definition is changed. Read-only views support the full feature set of  [selects](../../cds/cdl#as-select-from) to query data. 

::: tip Prefer simple views
Prefer creating simple views, each tailored to a specific use case, rather than a single complex view that tries to address multiple use cases simultaneously.
:::
```

:::

To [write data](#updatable-views) or [delete](#delete-via-view) through views, only use simple [projections](../../cds/cdl#as-projection-on). The CAP Java runtime attempts to resolve the CDS views to their underlying persistence entities, rewriting the statement and data accordingly, which is not supported for complex views.

For simple [projections](../../cds/cdl#as-projection-on), the generation of SQL views can be avoided by using [runtime views](#runtimeviews). This allows you to change the view definition without redeploying the database schema and is the prerequisite for lightweight extensibility via predefined extension fields.

::: tip Prefer simple views
Prefer creating multiple simple views, each tailored to a specific use case, rather than a single complex view that tries to address multiple use cases simultaneously.
:::
::: warning Avoid selecting paths over to-many Associations
Do not use [path expressions](../../cds/cql#path-expressions-in-all-other-clauses) over [*to-many associations*](../../cds/cdl#to-many-associations) in the select clause of CDS views. This blocks write operations and may cause performance issues due to record duplication on read.
:::

> ::: Avoid composition definitions in views

Author: ```suggestion
::: warning Avoid composition definitions in views
```

> ::: warning Avoid Composition Definitions in Views

Author: ```suggestion
::: Avoid composition definitions in views
```

Runtime views must be simple [projections](../../cds/cdl#as-projection-on), not using *aggregations*, *join*, *union* or *subqueries* in the *from* clause, but may have a *where* condition if they are only used to read. On write, the restrictions for [write through views](#updatable-views) apply in the same way as for standard CDS views. However, if a runtime view cannot be resolved, a fallback to database views is not possible, and the statement fails with an error.

CAP Java provides two modes for resolving runtime views during read operations: [cte](#rtview-cte) and [resolve](#rtview-resolve). 

::: details Changing the runtime view mode
To globally set the runtime view mode, use the property `cds.sql.runtimeView.mode` with value `cte` (the default) or `resolve` in the *application.yml*. To set the mode for a specific runtime view, annotate it with `@cds.java.runtimeView.mode: cte|resolve`.

To set the mode for a specific query, use a [hint](#hana-hints):

```Java
Select.from(BooksWithLowStock).hint("cds.sql.runtimeView.mode", "resolve");
```
:::

> The following section introduces both modes using the CDS model and query below:

Reviewer 2: ```suggestion
The next two sections introduce both modes using the following CDS model and query:
```

Prefer creating multiple simple views, each tailored to a specific use case, rather than a single complex view that tries to address multiple use cases simultaneously.
:::

The CDS compiler generates [DDL](../../guides/databases?impl-variant=java#generating-sql-ddl) files from your CDS model, including SQL views for the CDS views. These views are deployed to the [database](../cqn-services/persistence-services#database-support) and used by the CAP runtime to read data.

For *read-only* views, you can use the full feature set of [selects](../../cds/cdl#as-select-from), including *joins*, *unions*, and *aggregations*. However, such complex views are not writable and require a schema redeployment if the view definition is changed.

::: tip Indicate read-only
Use the `@readonly` annotation to indicate that a view or a view element is not writable.
:::

To [write data](#updatable-views) or [delete](#delete-via-view) through views, only use simple [projections](../../cds/cdl#as-projection-on). The CAP Java runtime attempts to resolve the CDS views to their underlying persistence entities, rewriting the statement and data accordingly, which is not supported for complex views.

> For simple projections, the generation of SQL views can be avoided by using [runtime views](#runtimeviews). This allows you to change the view definition without redeploying the database schema and is the prerequisite for lightweight extensibility via predefined extension fields.

Author: "we don't support unions or joins." - yes, therefor the "For simple projections", which don't allow unions and joins.

Nested projections can be used in cte mode but only with ucsn

::: warning Avoid selecting to-many Associations

> Do not use [*to-many associations*](../../cds/cdl#to-many-associations) in the select clause of CDS views. This blocks write operations and may cause performance issues due to record duplication on read.

Author: Yes, this was meant in the context of paths (see example in 275). I will make this clear

- The view is a simple [projection](../../cds/cdl#as-projection-on), not using *aggregations*, *join*, *union*, or *where*.
- The projection includes all *not null* elements (incl. keys), unless they have a default or generated value.
- The projection does not include [path expressions](../../cds/cql#path-expressions) using *to-many* associations.

If the runtime cannot resolve a view, the write operation is executed directly on the database view and the execution depends on the [database](../cqn-services/persistence-services#database-support) support.

For example, the following CDS model defines two resolvable views on the `Order` entity, which has a to-many *items* and a to-one *header* composition to a Header entity with a to-one customer *association*. In the projection, you can use path expressions as shown by *headerStatus* (writable) and *customerName* (read-only):

```cds
// CDS views supporting write operations
entity OrderView1 as projection on Order excluding { createdBy };

entity OrderView2 as projection on OrderView1 {
  key ID,
      header.status          as headerStatus,  // writable composition path
      header.customer.name   as customerName @readonly,

>       items                  as lineItems,        // writable composition

Author: 247 rather means to-many associations in paths. I will make this clear.

- [Path expressions](../../cds/cql#path-expressions) over compositions (*header.status*) are writable. For [Inserts](./query-api#insert), the view must expose all *not null* elements of the target entity and the data must include values for all of them. In the example above, the order header must have a generated key to support inserting new orders with a value for *headerStatus*.

    ::: warning Path Expressions over Associations
    Path expressions navigating *associations* (*header.customer.name*) are [not writable](#cascading-over-associations) by default. To avoid issues on write, annotate them with [@readonly](../../guides/providing-services#readonly).
    :::

### Delete through Views { #delete-via-view }

The CAP Java runtime attempts to resolve [Deletes](./query-api#delete) on CDS views to the underlying persistence entity definitions and the [write through views](#updatable-views) restrictions apply accordingly.

> If a view cannot be resolved, the delete operation is executed directly on the database view and the execution depends on the [database](../cqn-services/persistence-services#database-support) support.

Author: Probably rather the latter - we could skip the fallback

::: warning Annotate read-only views and elements
Use the `@readonly` annotation to indicate that a view or a view element is not writable, as this is not automatically detected by the CDS compiler.
:::

To [write data](#updatable-views) or [delete](#delete-via-view) through views, only use simple [projections](../../cds/cdl#as-projection-on). The CAP Java runtime attempts to resolve the CDS views to their underlying persistence entities, rewriting the statement and data accordingly, which is not supported for complex views.

For simple [projections](../../cds/cdl#as-projection-on), the generation of SQL views can be avoided by using [runtime views](#runtimeviews). This allows you to change the view definition without redeploying the database schema and is the prerequisite for lightweight extensibility via predefined extension fields.

::: tip Prefer simple views

> Prefer creating multiple simple views, each tailored to a specific use case, rather than a single complex view that tries to address multiple use cases simultaneously.

Author: -> https://github.com/cap-js/docs/pull/1957
